pkg/util/chunkinfologger/chunkinfologger.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package chunkinfologger
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"strings"
+
+	"github.com/go-kit/log"
+
+	"github.com/grafana/mimir/pkg/ingester/client"
+	"github.com/grafana/mimir/pkg/querier/api"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// Allow for 10% overhead from string quotes when logger dumps the JSON.
+const maxSize = (64 * 1024 * 90) / 100
+
+const ChunkInfoLoggingHeader = "X-Mimir-Chunk-Info-Logger"
+const ChunkInfoLoggingEnabled = "true"
+
+func ChunkInfoLoggingHeaderValidator(v string) error {
+	if v == ChunkInfoLoggingEnabled {
+		return nil
+	}
+	return fmt.Errorf("must be exactly 'true' or not set")
+}
+
+func EnableChunkInfoLoggingFromContext(ctx context.Context) bool {
+	v, ok := api.HeaderOptionFromContext(ctx, ChunkInfoLoggingHeader)
+	return ok && v == ChunkInfoLoggingEnabled
+}
+
+type ChunkInfoLogger struct {
+	chunkInfo strings.Builder
+	msg       string
+	traceID   string
+	logger    log.Logger
+
+	// sourceCount is the number of sources in the current series to track if we need to add a comma.
+	sourceCount int
+}
+
+func NewChunkInfoLogger(msg, traceID string, logger log.Logger) *ChunkInfoLogger {
+	return &ChunkInfoLogger{
+		chunkInfo: strings.Builder{},
+		msg:       msg,
+		traceID:   traceID,
+		logger:    logger,
+	}
+}
+
+func (c *ChunkInfoLogger) SetMsg(msg string) {
+	c.msg = msg
+}
+
+func (c *ChunkInfoLogger) StartSeries(seriesID string) {
+	c.sourceCount = 0
+	if c.chunkInfo.Len() > 0 {
+		c.chunkInfo.WriteString(",\"") // next series
```

Comment:
[nit] Might be clearer to use raw strings in cases like this in this file

```suggestion
		c.chunkInfo.WriteString(`,"`) // next series
```